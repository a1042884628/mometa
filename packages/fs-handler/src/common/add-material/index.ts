import { NodePath, PluginObj, Visitor, transformAsync, transformSync } from '@babel/core'
import groupBy from 'lodash.groupby'
import { InsertNodePreload, OpType, Point, RequestData } from '../../index'
import { createLineContentsByContent, LineContents, Range } from '../../utils/line-contents'

const sideEffectImportVisitor = {
  ImportDeclaration(path, state) {
    if (!path.get('specifiers')?.length) {
      state.result.set(path.get('source').node.value, path)
    }
    state.startImportLoc = state.startImportLoc ?? path.node.loc
    state.endImportLoc = path.node.loc
    path.skip()
  }
} as Visitor<SideEffectImportVisitorState>

type SideEffectImportVisitorState = { result: Map<string, NodePath>; startImportLoc: Range; endImportLoc: Range }

type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>
}

type Config = {
  esModule?: boolean
}

const parserOptsPlugins: any[] = [
  'jsx',
  'asyncDoExpressions',
  'asyncGenerators',
  'bigInt',
  'classPrivateMethods',
  'classPrivateProperties',
  'classProperties',
  'classStaticBlock',
  'decimal',
  'decorators-legacy',
  'doExpressions',
  'dynamicImport',
  'exportDefaultFrom',
  'exportNamespaceFrom',
  'functionBind',
  'functionSent',
  'importMeta',
  'logicalAssignment',
  'importAssertions',
  'moduleBlocks',
  'moduleStringNames',
  'nullishCoalescingOperator',
  'numericSeparator',
  'objectRestSpread',
  'optionalCatchBinding',
  'optionalChaining',
  'partialApplication',
  'placeholders',
  'privateIn',
  'throwExpressions',
  'topLevelAwait',
  'typescript'
]

const replaceCodePlugin = ({ types: t }) => {
  const handleIdentifier = (path: NodePath<any>, opts: any, rawCode: string) => {
    opts.output.code = opts.output.code ?? rawCode
    opts.state = opts.state ?? {
      deltaOffset: 0
    }

    const prevCode = opts.output.code
    if (opts.input.validData && Reflect.has(opts.input.validData, path.node.name)) {
      const oldVal = path.node.name
      const newVal = opts.input.validData[path.node.name]
      const deltaOffset = opts.state.deltaOffset

      opts.output.code =
        prevCode.slice(0, path.node.start + deltaOffset) +
        opts.input.validData[path.node.name] +
        prevCode.slice(path.node.end + deltaOffset)

      opts.state.deltaOffset += newVal.length - oldVal.length
    }
  }

  return {
    visitor: {
      Program(p, { opts }) {},
      JSXIdentifier(path, { opts }) {
        // @ts-ignore
        handleIdentifier(path, opts, this.file.code)
      },
      Identifier(path, { opts }) {
        // @ts-ignore
        handleIdentifier(path, opts, this.file.code)
      }
    }
  } as PluginObj<{
    opts: {
      input: {
        validData: Record<string, string>
      }
      output: {
        code?: string
      }
    }
  }>
}

const importPlugin = ({ types: t }) =>
  ({
    visitor: {
      Program(path, state) {
        const { material, pos, ops, config } = state.opts
        const { esModule = true } = config || {}

        // @ts-ignore
        const rawCode = this.file.code
        // @ts-ignore
        const lineContents = createLineContentsByContent(rawCode, { filename: this.filename })
        const getText = ({ start, end }) => {
          const arr = lineContents.locateByRange({
            start,
            end
          })
          return arr
            .map(({ lineNumber, line, start, end }) => {
              return line.toString()
            })
            .join('\n')
        }

        const sideEffectImportVisitorState = {
          result: new Map(),
          endImportLoc: { end: { line: 1, column: 0 }, start: { line: 1, column: 0 } }
        } as SideEffectImportVisitorState
        path.traverse(sideEffectImportVisitor, sideEffectImportVisitorState)
        if (material.sideEffectDependencies) {
          material.sideEffectDependencies.forEach((depName) => {
            if (!sideEffectImportVisitorState.result.has(depName)) {
              ops.insertDeps.push({
                type: OpType.INSERT_NODE,
                preload: {
                  to: sideEffectImportVisitorState.endImportLoc.end,
                  data: {
                    newText: esModule ? `;import ${JSON.stringify(depName)};` : `;require(${JSON.stringify(depName)});`
                  }
                }
              })
            }
          })
        }

        if (material.dependencies && Object.keys(material.dependencies)?.length) {
          const map = new Map<string, { path?: NodePath; name: any }>()
          const namedUnionMap = new Map<string, NodePath[]>()
          for (const [name, binding] of Object.entries(path.scope.bindings)) {
            if (binding.kind === 'module') {
              const importDecPath = binding.path.findParent((x) => x.isImportDeclaration())
              if (importDecPath && (importDecPath.get('source') as any).node?.value) {
                const source = (importDecPath.get('source') as any).node?.value
                // @ts-ignore
                const imported = binding.path.node?.imported as any
                if (binding.path.isImportSpecifier() && (imported?.name ?? imported?.value) !== 'default') {
                  map.set(`${source}#named:${imported.name ?? imported.value}`, {
                    name: binding.path.node.local.name,
                    path: importDecPath
                  })
                  namedUnionMap.set(
                    source,
                    namedUnionMap.get(source) ? namedUnionMap.get(source).concat(importDecPath) : [importDecPath]
                  )
                } else if (
                  binding.path.isImportDefaultSpecifier() ||
                  (imported?.name ?? imported?.value) === 'default'
                ) {
                  map.set(`${source}#default`, {
                    // @ts-ignore
                    name: binding.path.node.local.name,
                    path: importDecPath
                  })
                } else if (binding.path.isImportNamespaceSpecifier()) {
                  map.set(`${source}#namespace`, {
                    name: binding.path.node.local.name,
                    path: importDecPath
                  })
                }
              }
            }
          }

          const namedCache = new Map()
          const getValidName = (name = 'Unknown') => {
            name = name.replace(/[\/@]/g, '$')
            // @ts-ignore
            while (true === path.scope?.references?.[name] || namedCache.has(name)) {
              name = `_${name}`
            }
            namedCache.set(name, true)
            return name
          }

          const tplData = {}
          const list = Object.keys(material.dependencies).map((tplName) => ({
            ...material.dependencies[tplName],
            tplName
          }))
          const sourceMap = groupBy(list, 'source')
          for (const [source, dataList] of Object.entries(sourceMap)) {
            const modeMap = groupBy(dataList, (d) =>
              d.imported === 'default' || d.mode === 'default' ? 'default' : d.mode
            )
            for (const [mode, dataList] of Object.entries(modeMap)) {
              // @ts-ignore
              if (!dataList.length) return
              const cachedKey = `${source}#${mode}`
              const cached = map.get(cachedKey)
              if (mode === 'default') {
                if (cached?.name) {
                  tplData[dataList[0].tplName] = cached?.name
                } else {
                  const name = getValidName(dataList[0].local ?? dataList[0].imported ?? source)
                  map.set(cachedKey, { name })
                  tplData[dataList[0].tplName] = name
                  ops.insertDeps.push({
                    type: OpType.INSERT_NODE,
                    preload: {
                      to: sideEffectImportVisitorState.endImportLoc.end,
                      data: {
                        newText: esModule
                          ? `;import ${name} from ${JSON.stringify(source)};`
                          : `;var ${name} = (function () {
                        var mod = require(${JSON.stringify(source)});
                        return mod.__esModule ? mod.default : mod
                        })();`
                      }
                    }
                  })
                }
              } else if (mode === 'namespace') {
                if (cached?.name) {
                  tplData[dataList[0].tplName] = cached?.name
                } else {
                  const name = getValidName(dataList[0].local ?? dataList[0].imported ?? source)
                  map.set(cachedKey, { name })
                  tplData[dataList[0].tplName] = name
                  ops.insertDeps.push({
                    type: OpType.INSERT_NODE,
                    preload: {
                      to: sideEffectImportVisitorState.endImportLoc.end,
                      data: {
                        newText: esModule
                          ? `;import * as ${name} from ${JSON.stringify(source)};`
                          : `;var ${name} = require(${JSON.stringify(source)});`
                      }
                    }
                  })
                }
              } else if (mode === 'named') {
                const newNamedPathSet = new Set<NodePath>()
                const newNamedSet = new Set<{ local: string; imported: string }>()
                ;(dataList as any).forEach((data) => {
                  const key = `${cachedKey}:${data.imported}`
                  const cached = map.get(key)
                  if (cached?.name) {
                    tplData[data.tplName] = cached?.name
                  } else {
                    let importPath
                    // insert into the last importPath
                    if (namedUnionMap.get(source)?.length) {
                      importPath = namedUnionMap.get(source)[namedUnionMap.get(source)?.length - 1]
                    }
                    const name = getValidName(data.local ?? data.imported)

                    if (importPath) {
                      importPath.node.specifiers.push(
                        t.importSpecifier(t.identifier(name), t.identifier(data.imported))
                      )
                    }

                    map.set(key, { name, path: importPath })
                    tplData[data.tplName] = name
                    if (importPath) {
                      newNamedPathSet.add(importPath)
                    } else {
                      newNamedSet.add({
                        local: name,
                        imported: data.imported
                      })
                    }
                  }
                })

                // 在已有的 import 中进行替换
                if (newNamedPathSet.size) {
                  for (const importPath of newNamedPathSet.values()) {
                    let cjsCodes = []
                    // @ts-ignore
                    importPath.node?.specifiers?.forEach((x) => {
                      if (x.local.name === x.imported.name) {
                        cjsCodes.push(x.local.name)
                      } else {
                        cjsCodes.push(`${x.imported.name}: ${x.local.name}`)
                      }
                    })
                    ops.insertDeps.push({
                      type: OpType.REPLACE_NODE,
                      preload: {
                        ...importPath.node.loc,
                        text: getText(importPath.node.loc),
                        data: {
                          newText: (esModule
                            ? `;${importPath.toString()};`
                            : // @ts-ignore
                              `;var { ${cjsCodes.join(',')} } = require(${importPath.node.source.value});`
                          ).replace(/;+$/, ';')
                        }
                      }
                    })
                  }
                }
                if (newNamedSet.size) {
                  const stringifiedSource = JSON.stringify(source)

                  const values = []
                  newNamedSet.forEach(({ local, imported }) => {
                    if (local === imported) {
                      values.push(local)
                      return
                    }
                    if (esModule) {
                      values.push(`${imported} as ${local}`)
                    } else {
                      values.push(`${imported} : ${local}`)
                    }
                  })

                  ops.insertDeps.push({
                    type: OpType.INSERT_NODE,
                    preload: {
                      to: sideEffectImportVisitorState.endImportLoc.end,
                      data: {
                        newText: esModule
                          ? `;import { ${values.join(',')} } from ${stringifiedSource};`
                          : `;var { ${values.join(',')} } = require(${stringifiedSource});`
                      }
                    }
                  })
                }
              }
            }
          }

          let newCode: string = material.code
          if (newCode) {
            const validData = Object.create(null)
            Object.keys(tplData).forEach((tplName) => {
              if (newCode && tplName !== tplData[tplName]) {
                validData[tplName] = tplData[tplName]
              }
            })

            if (Object.keys(validData)?.length) {
              const output: any = {}
              transformSync(newCode, {
                parserOpts: {
                  plugins: parserOptsPlugins
                },
                babelrc: false,
                plugins: [[replaceCodePlugin, { input: { validData }, output }]],
                ast: false,
                code: false
              })
              newCode = output.code ?? newCode
            }
          }

          ops.insertCode = {
            type: OpType.INSERT_NODE,
            preload: {
              to: pos,
              data: {
                newText: newCode
              }
            }
          }
        }

        path.stop()
      }
    }
  } as PluginObj<{
    opts: {
      config?: Config
      pos: Point
      material: InsertNodePreload['data']['material']
      lineContents: LineContents
      ops: {
        insertDeps: Array<DeepPartial<RequestData>>
        insertCode: DeepPartial<RequestData>
      }
    }
  }>)

export async function getAddMaterialOps(
  lineContents: LineContents,
  pos: Point,
  material: InsertNodePreload['data']['material'],
  config?: Config
) {
  const ops: {
    insertDeps: Array<RequestData>
    insertCode?: RequestData
  } = {
    insertDeps: [],
    insertCode: null
  }

  if (material?.dependencies) {
    const dependencies = material?.dependencies
    material = {
      ...material,
      dependencies: Object.keys(dependencies).reduce((acc, name) => {
        const vo = dependencies[name]
        acc[name] = {
          ...vo,
          local: vo.local ?? name,
          imported: vo.mode === 'named' ? vo.imported ?? name : vo.imported
        }
        return acc
      }, {})
    }
  }

  await transformAsync(lineContents.toString(false), {
    filename: lineContents.options?.filename,
    parserOpts: {
      plugins: parserOptsPlugins
    },
    babelrc: false,
    plugins: [[importPlugin, { lineContents, material, pos, ops, config }]],
    ast: false,
    code: false
  })
  ;(ops.insertDeps.concat(ops.insertCode) as any).forEach((op) => {
    if (op?.preload && lineContents?.options) {
      op.preload.filename = lineContents?.options.filename
    }
  })

  return ops
}
