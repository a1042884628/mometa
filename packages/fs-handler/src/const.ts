export enum OpType {
  DEL,
  REPLACE_NODE,
  MOVE_NODE,
  INSERT_NODE
}
