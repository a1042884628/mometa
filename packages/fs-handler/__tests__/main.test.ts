import { createFsHandler, OpType, commonMiddlewares, createLineContentsByContent } from '../src'
import * as fs from 'fs'
import { fixture, readContent } from './helper'
import { getAddMaterialOps } from '../src/common/add-material'

const mockFs = {
  readFile: fs.readFile,
  writeFile: fs.writeFile
}

describe('fsHandler', function () {
  const raw = mockFs.writeFile
  let result
  let fsHandler
  beforeEach(() => {
    fsHandler = createFsHandler({
      fs: mockFs,
      middlewares: commonMiddlewares()
    })
    // @ts-ignore
    Object.defineProperty(mockFs, 'writeFile', {
      value: (path, data, cb) => {
        result = data
        cb(null)
      },
      configurable: true
    })
  })
  afterEach(() => {
    mockFs.writeFile = raw
    result = null
  })

  it('spec case DEL', async function () {
    await fsHandler({
      type: OpType.DEL,
      preload: {
        name: 'p',
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        start: { line: 18, column: 10 },
        end: { line: 18, column: 21 }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('spec case REPLACE_NODE', async function () {
    await fsHandler({
      type: OpType.REPLACE_NODE,
      preload: {
        data: {
          newText: '<span>ppp</span>'
        },
        name: 'p',
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        start: { line: 18, column: 10 },
        end: { line: 18, column: 21 }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('nest REPLACE_NODE', async function () {
    await fsHandler({
      type: OpType.REPLACE_NODE,
      preload: {
        start: { line: 17, column: 10 },
        end: { line: 20, column: 14 },
        filename: fixture('nested.tsx'),
        name: 'p',
        text: '<p>\n            nested\n            <strong>hahahax</strong>\n          </p>',
        data: {
          newText: '<p>\n            nested 2\n            <strong>hahahax</strong>\n          </p>'
        }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('MOVE_NODE', async function () {
    await fsHandler({
      type: OpType.MOVE_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        start: { line: 18, column: 10 },
        end: { line: 18, column: 21 },
        data: {
          to: {
            line: 17,
            column: 10
          }
        }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('INSERT_NODE', async function () {
    await fsHandler({
      type: OpType.INSERT_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        to: { line: 18, column: 10 },
        data: {
          newText: '<p className="empty-2"></p>'
        }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('INSERT_NODE material', async function () {
    await fsHandler({
      type: OpType.INSERT_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        to: { line: 18, column: 10 },
        data: {
          material: {
            code: '<Button type="default">按钮</Button>',
            dependencies: {
              Button: {
                source: 'antd',
                mode: 'named'
              }
            }
          }
        }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('INSERT_NODE material existed name', async function () {
    await fsHandler({
      type: OpType.INSERT_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        to: { line: 18, column: 10 },
        data: {
          material: {
            code: '<Tabs><Tabs.Panel></Tabs.Panel></Tabs>',
            dependencies: {
              Tabs: {
                source: 'antd',
                mode: 'named'
              }
            }
          }
        }
      }
    })
    expect(result).toMatchSnapshot()
  })

  it('INSERT_NODE wrap', async function () {
    await fsHandler({
      type: OpType.INSERT_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        to: { line: 18, column: 10 },
        data: {
          wrap: {
            anotherTo: { line: 17, column: 10 },
            startStr: '<>',
            endStr: '</>'
          },
          newText: '<p className="empty-2"></p>'
        }
      }
    })
    expect(result).toMatchSnapshot()

    await fsHandler({
      type: OpType.INSERT_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        to: { line: 17, column: 10 },
        data: {
          wrap: {
            anotherTo: { line: 18, column: 10 },
            startStr: '<>',
            endStr: '</>'
          },
          newText: '<p className="empty-2"></p>'
        }
      }
    })
    expect(result).toMatchSnapshot()

    await fsHandler({
      type: OpType.INSERT_NODE,
      preload: {
        text: '<p>单独 p</p>',
        filename: fixture('react-comp.tsx'),
        to: { line: 18, column: 10 },
        data: {
          wrap: {
            anotherTo: { line: 18, column: 10 },
            startStr: '<>',
            endStr: '</>'
          },
          newText: '<p className="empty-2"></p>'
        }
      }
    })
    expect(result).toMatchSnapshot()
  })

  describe('getAddMaterialOps', function () {
    it('spec', async function () {
      const lineContents = createLineContentsByContent(readContent('simple.tsx'), { filename: '/a.js' })

      const ops = await getAddMaterialOps(
        lineContents,
        {
          line: 6,
          column: 3
        },
        {
          code: `
    <>
      <Antd />
      <Button type="default">按钮</Button>
    </>
    `.trim(),
          dependencies: {
            Button: {
              source: 'antd',
              mode: 'named'
            },
            Antd: {
              source: 'antd',
              mode: 'default'
            }
          } as any,
          sideEffectDependencies: ['antd/lib/button/style', 'antd/lib/button/style/css']
        }
      )
      expect(ops).toMatchSnapshot()
    })

    it('spec cjs', async function () {
      const lineContents = createLineContentsByContent(readContent('simple.tsx'), { filename: '/a.js' })

      const ops = await getAddMaterialOps(
        lineContents,
        {
          line: 6,
          column: 3
        },
        {
          code: `
    <>
      <Antd />
      <Button type="default">按钮</Button>
    </>
    `.trim(),
          dependencies: {
            Button: {
              source: 'antd',
              mode: 'named'
            },
            Antd: {
              source: 'antd',
              mode: 'default'
            }
          } as any,
          sideEffectDependencies: ['antd/lib/button/style', 'antd/lib/button/style/css']
        },
        { esModule: false }
      )
      expect(ops).toMatchSnapshot()
    })
    it('spec fsHandler', async function () {
      await fsHandler({
        type: OpType.INSERT_NODE,
        preload: {
          to: {
            line: 4,
            column: 18
          },
          filename: fixture('simple.tsx'),
          data: {
            material: {
              code: `
    <>
      <Antd />
      <Button type="default">按钮</Button>
    </>
    `.trim(),
              dependencies: {
                Button: {
                  source: 'antd',
                  mode: 'named'
                },
                Antd: {
                  source: 'antd',
                  mode: 'default'
                }
              } as any,
              sideEffectDependencies: ['antd/lib/button/style', 'antd/lib/button/style/css']
            }
          }
        }
      })

      expect(result).toMatchSnapshot()
    })

    it('fsHandler', async function () {
      await fsHandler({
        type: OpType.INSERT_NODE,
        preload: {
          filename: fixture('react-comp.tsx'),
          to: { line: 18, column: 10 },
          data: {
            wrap: {
              anotherTo: { line: 18, column: 10 },
              startStr: '<>',
              endStr: '</>'
            },
            material: {
              code: `
    <>
      <Antd />
      <Button type="default">按钮</Button>
    </>
    `.trim(),
              dependencies: {
                Button: {
                  source: 'antd',
                  mode: 'named'
                },
                Antd: {
                  source: 'antd',
                  mode: 'default'
                }
              } as any,
              sideEffectDependencies: ['antd/lib/button/style', 'antd/lib/button/style/css']
            }
          }
        }
      })
      expect(result).toMatchSnapshot()
    })
  })
})
