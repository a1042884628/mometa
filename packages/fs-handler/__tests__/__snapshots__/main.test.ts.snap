// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fsHandler INSERT_NODE 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <p className=\\"empty-2\\"></p>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler INSERT_NODE material 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'
import { Button } from 'antd'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <Button type=\\"default\\">按钮</Button>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler INSERT_NODE material existed name 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'
import { Tabs as _Tabs } from 'antd'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <_Tabs>
            <_Tabs.Panel></_Tabs.Panel>
          </_Tabs>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler INSERT_NODE wrap 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <>
            <p className=\\"empty\\"></p>
            <p className=\\"empty-2\\"></p>
          </>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler INSERT_NODE wrap 2`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <>
            <p className=\\"empty-2\\"></p>
            <p className=\\"empty\\"></p>
          </>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler INSERT_NODE wrap 3`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <>
            <p className=\\"empty-2\\"></p>
          </>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler MOVE_NODE 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p>单独 p</p>
          <p className=\\"empty\\"></p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler getAddMaterialOps fsHandler 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'
import Antd from 'antd'
import { Button } from 'antd'
import 'antd/lib/button/style/css'
import 'antd/lib/button/style'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <>
            <>
              <Antd />
              <Button type=\\"default\\">按钮</Button>
            </>
          </>
          <p>单独 p</p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler getAddMaterialOps spec 1`] = `
Object {
  "insertCode": Object {
    "preload": Object {
      "data": Object {
        "newText": "<>
      <Antd />
      <Button type=\\"default\\">按钮</Button>
    </>",
      },
      "filename": "/a.js",
      "to": Object {
        "column": 3,
        "line": 6,
      },
    },
    "type": 3,
  },
  "insertDeps": Array [
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";import \\"antd/lib/button/style\\";",
        },
        "filename": "/a.js",
        "to": Position {
          "column": 20,
          "line": 2,
        },
      },
      "type": 3,
    },
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";import \\"antd/lib/button/style/css\\";",
        },
        "filename": "/a.js",
        "to": Position {
          "column": 20,
          "line": 2,
        },
      },
      "type": 3,
    },
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";import { Input, Button } from 'antd';",
        },
        "end": Position {
          "column": 28,
          "line": 1,
        },
        "filename": "/a.js",
        "identifierName": undefined,
        "start": Position {
          "column": 0,
          "line": 1,
        },
        "text": "import { Input } from 'antd'",
      },
      "type": 1,
    },
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";import Antd from \\"antd\\";",
        },
        "filename": "/a.js",
        "to": Position {
          "column": 20,
          "line": 2,
        },
      },
      "type": 3,
    },
  ],
}
`;

exports[`fsHandler getAddMaterialOps spec cjs 1`] = `
Object {
  "insertCode": Object {
    "preload": Object {
      "data": Object {
        "newText": "<>
      <Antd />
      <Button type=\\"default\\">按钮</Button>
    </>",
      },
      "filename": "/a.js",
      "to": Object {
        "column": 3,
        "line": 6,
      },
    },
    "type": 3,
  },
  "insertDeps": Array [
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";require(\\"antd/lib/button/style\\");",
        },
        "filename": "/a.js",
        "to": Position {
          "column": 20,
          "line": 2,
        },
      },
      "type": 3,
    },
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";require(\\"antd/lib/button/style/css\\");",
        },
        "filename": "/a.js",
        "to": Position {
          "column": 20,
          "line": 2,
        },
      },
      "type": 3,
    },
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";var { Input,Button } = require(antd);",
        },
        "end": Position {
          "column": 28,
          "line": 1,
        },
        "filename": "/a.js",
        "identifierName": undefined,
        "start": Position {
          "column": 0,
          "line": 1,
        },
        "text": "import { Input } from 'antd'",
      },
      "type": 1,
    },
    Object {
      "preload": Object {
        "data": Object {
          "newText": ";var Antd = (function () {
                        var mod = require(\\"antd\\");
                        return mod.__esModule ? mod.default : mod
                        })();",
        },
        "filename": "/a.js",
        "to": Position {
          "column": 20,
          "line": 2,
        },
      },
      "type": 3,
    },
  ],
}
`;

exports[`fsHandler getAddMaterialOps spec fsHandler 1`] = `
"import { Input, Button } from 'antd'
import 'side-effect'
import Antd from 'antd'
import 'antd/lib/button/style/css'
import 'antd/lib/button/style'

const elem = (
  <div>
    <>
      <Antd />
      <Button type=\\"default\\">按钮</Button>
    </>
    abc
  </div>
)

export { Button } from 'antd'"
`;

exports[`fsHandler nest REPLACE_NODE 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World👌</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <p>simple 66xxxxxasdasdas6</p>
          <p>
            nested 2<strong>hahahax</strong>
          </p>
          {array.map((x, i) => (
            <p key={i}>物料_{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler spec case DEL 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;

exports[`fsHandler spec case REPLACE_NODE 1`] = `
"import React, { StrictMode } from 'react'
import Tabs from 'antd/es/tabs'
import 'antd/es/tabs/style/index.css'

type Props = {
  x: string
}

const array = new Array(100).fill(1)

export default function App(props: Props) {
  return (
    <div>
      <h1 title={'abc'}>Hello World</h1>
      <Tabs>
        <Tabs.TabPane key={'tool'} tab={'物料'}>
          <p className=\\"empty\\"></p>
          <span>ppp</span>
          {array.map((x, i) => (
            <p key={i}>物料__{i}</p>
          ))}
        </Tabs.TabPane>
        <Tabs.TabPane key={'attr'} tab={'属性'}></Tabs.TabPane>
      </Tabs>
    </div>
  )
}"
`;
